src/providers/solana/rateLimitedRpcFactory.ts

```diff
@@ -0,0 +1,101 @@
+import { createDefaultRpcTransport, RpcResponse, RpcTransport } from "@solana/web3.js";
+import { QueueObject, queue } from "async";
+import winston, { Logger } from "winston";
+import { SolanaClusterRpcFactory } from "./baseRpcFactories";
+import { SolanaRateLimitTask } from "./utils";
+import { getOriginFromURL } from "../../utils";
+
+// This factory is a very small addition to the createDefaultRpcTransport that ensures that no more than maxConcurrency
+// requests are ever in flight. It uses the async/queue library to manage this.
+export class RateLimitedSolanaRpcFactory extends SolanaClusterRpcFactory {
+  // The queue object that manages the tasks.
+  private queue: QueueObject<SolanaRateLimitTask>;
+
+  // Holds the underlying transport that the rate limiter wraps.
+  private readonly baseTransport: RpcTransport;
+
+  // Takes the same arguments as the createDefaultRpcTransport, but it has an additional parameters to control
+  // concurrency and logging at the beginning of the list.
+  constructor(
+    maxConcurrency: number,
+    readonly pctRpcCallsLogged: number,
+    readonly logger: Logger = winston.createLogger({
+      transports: [new winston.transports.Console()],
+    }),
+    chainId: number,
+    ...rpcTransportParams: Parameters<typeof createDefaultRpcTransport>
+  ) {
+    super(chainId, rpcTransportParams[0].url);
+    this.baseTransport = createDefaultRpcTransport(...rpcTransportParams);
+
+    // This sets up the queue. Each task is executed by forwarding the RPC request to the underlying base transport.
+    // This queue sends out requests concurrently, but stops once the concurrency limit is reached. The maxConcurrency
+    // is configured here.
+    this.queue = queue(async ({ rpcArgs, resolve, reject }: SolanaRateLimitTask, callback: () => void) => {
```

Comment:
This queuing logic looks almost identical to the EVM version, is there any way to refactor and re-use code?